File: pl1/tutorial1-threading/java/src/main/java/PC_wrong.java
class Q {
  int n;

  synchronized int get() {
    System.out.println("Got: " + n);
    notify();
    try {
      wait();
    } catch (InterruptedException e) {
      // TODO: handle exception
      System.out.println("algo correu mal\n");
    }
    return n;
  }

  synchronized void put(int n) {
    this.n = n;
    try {
      wait();
    } catch (InterruptedException e) {
      // TODO: handle exception
      System.out.println("algo correu mal\n");
    }
    System.out.println("Put: " + n);
    notify();
  }
}

class Producer implements Runnable {
  Q q;

  Producer(Q q) {
    this.q = q;
    new Thread(this, "Producer").start();
  }

  public void run() {
    int i = 0;
    while (i < 100) {
      q.put(i++);
    }
  }
}

class Consumer implements Runnable {
  Q q;

  Consumer(Q q) {
    this.q = q;
    new Thread(this, "Consumer").start();
  }

  public void run() {
    int i = 0;
    while (i < 100) {
      q.get();
      i++;
    }
  }
}

class PC_wrong {
  public static void main(String args[]) {
    Q q = new Q();
    new Producer(q);
    new Consumer(q);
  }
}

File: pl1/tutorial1-threading/python/pc_wrong.py
import threading

class Q:
    def __init__(self):
        self.n = 0
        self.lock = threading.Lock()

    def get(self):
        with self.lock:
            print("Got:", self.n)
            return self.n

    def put(self, n):
        with self.lock:
            self.n = n
            print("Put:", n)

class Producer(threading.Thread):
    def __init__(self, q):
        threading.Thread.__init__(self)
        self.q = q
        self.start()

    def run(self):
        i = 0
        while i < 100:
            self.q.put(i)
            i += 1

class Consumer(threading.Thread):
    def __init__(self, q):
        threading.Thread.__init__(self)
        self.q = q
        self.start()

    def run(self):
        i = 0
        while i < 100:
            self.q.get()
            i += 1

if __name__ == "__main__":
    q = Q()
    Producer(q)
    Consumer(q)
